internal/expr/parser.go

```diff
@@ -716,30 +748,148 @@ func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
+	}
+	if ma.memberName == "*" {
+		cp.restore()
+		return nil, false, errorAt(fmt.Errorf("invalid asterisk placement in input %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	}
+	return &memberInputExpr{ma: ma, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseAsteriskInputExpr parses an INSERT statement input expression where
+// SQLair generates the columns.
+// It is of the form "(*) VALUES ($Type.*, $Type.member,...)".
+func (p *Parser) parseAsteriskInputExpr() (*asteriskInputExpr, bool, error) {
+	cp := p.save()
+	if !p.skipByte('(') {
+		return nil, false, nil
+	}
+	p.skipBlanks()
+	if !p.skipByte('*') {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+	if !p.skipByte(')') {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+	if !p.skipString("VALUES") {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+
+	sources, ok, err := parseList(p, (*Parser).parseInputMemberAccessor)
+	if err != nil {
+		cp.restore()
+		return nil, false, err
+	} else if !ok {
+		// Check for types with missing parentheses.
+		if _, ok, _ := p.parseInputMemberAccessor(); ok {
+			err = errorAt(fmt.Errorf(`missing parentheses around types after "VALUES"`), cp.lineNum, cp.colNum(), p.input)
 		}
-		return &memberInputExpr{ma: ma, raw: p.input[cp.pos:p.pos]}, true, nil
-	} else if err != nil {
+		cp.restore()
 		return nil, false, err
 	}
 
-	cp.restore()
+	return &asteriskInputExpr{sources: sources, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseColumnsInputExpr parses an INSERT statement input expression where the
+// user specifies the columns to insert from a single type.
+// It is of the form "(col1, col2,...) VALUES ($Type.*)".
+func (p *Parser) parseColumnsInputExpr() (*columnsInputExpr, bool, error) {
+	cp := p.save()
+
+	columns, paren, ok := p.parseColumns()
+	if !(ok && paren) {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+	if !p.skipString("VALUES") {
+		cp.restore()
+		return nil, false, nil
+	}
+	p.skipBlanks()
+
+	parenCol := p.colNum()
+	parenLine := p.lineNum
+	// Ignore errors and leave them to be handled by
+	// parseMemberInputExpr later.
+	sources, ok, _ := parseList(p, (*Parser).parseInputMemberAccessor)
+	if !ok {
+		// Check for types with missing parentheses.
+		if _, ok, _ := p.parseTypeAndMember(); ok {
+			cp.restore()
+			return nil, false, errorAt(fmt.Errorf(`missing parentheses around types after "VALUES"`), parenLine, parenCol, p.input)
+		}
+		cp.restore()
+		return nil, false, nil
+	}
+	if starCountTypes(sources) == 0 {
+		// If there are no asterisk accessors leave the accessors then leave
+		// the accessors to be handled elsewhere.
+		cp.restore()
+		return nil, false, nil
+	}
+
+	return &columnsInputExpr{columns: columns, sources: sources, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseInputExpr parses all forms of input expressions, that is, expressions
+// containing a "$".
+func (p *Parser) parseInputExpr() (expression, bool, error) {
```

Comment:
I would tend to locate this above the methods it calls.

Reply:
You could also do a loop over over the methods as first class funcs and shorten this.

Reply:
Moved and changed.